### PR TITLE
Avoid freezing of EditorPresenter and HomePresenter and their deps

### PR DESCRIPTION
**Pull request by arkivanov**
WIP

### CODE REVIEW

**Comment by arkivanov** [comment]
@saket Tests for iOS do not run. There is no test task for iOS, like `iosTest`. Looks like some additional configuration is required, but I'm not really familiar with it.

**Comment by tobiasheine** [comment]
You could copy the configuration for such a task from here: https://github.com/tobiasheine/multiplatform-movies/blob/master/frontend/build.gradle#L134

**Comment by msfjarvis** [comment]
Kotlin's MPP sample seems to do a similar thing, going to take this for a spin and see if I can make it work.

https://github.com/Kotlin/mpp-example/blob/2ded37d623b6219f531e38b4a5546eb3b664792a/greeting/build.gradle#L83

**Comment by saket** [comment]
Fwiw I was also trying to add iOS tests, but I'm hitting many roadbumps. My progress can be found in this branch: https://github.com/saket/Press/commits/saket/dec11/ios-tests

**Comment by arkivanov** [comment]
Blocked by not running iOS tests

**Comment by saket** [comment]
I'm working on this. I learned today that libraries compiled with old versions of Kotlin Native don't play very nice, so I need to find a way to bring everything to the same version.

**Comment by arkivanov** [comment]
> I'm working on this. I learned today that libraries compiled with old versions of Kotlin Native don't play very nice, so I need to find a way to bring everything to the same version.

Yeah, more dependencies you have more time you wait to migrate to a new Kotlin version (you wait for all your deps to be updated).

**Comment by saket** [comment]
Tracking this issue here: https://github.com/saket/press/pull/22

File: shared/src/commonTest/kotlin/me/saket/press/shared/note/FakeNoteRepository.kt [hunk 1]
```diff
@@ -29,23 +33,23 @@ class FakeNoteRepository : NoteRepository {
         includeEmptyNotes -> savedNotes
         else -> savedNotes.filter { it.content.isNotEmpty() }
       }
-    }
+    }.observeOn(scheduler)
```
Comment (saket) [inline shared/src/commonTest/kotlin/me/saket/press/shared/note/FakeNoteRepository.kt:44]: What's the reason for explicitly specifying the schedulers here in fake code when `observableFromFunction()` will run on the thread it was subscribed on anyway?

**Comment by saket** [comment]
@arkivanov you should now be able to run iOS tests using `./gradlew iosTest`. I'm seeing quite a lot of errors related to freezing, especially in `EditorPresenterTest`.

Update: I was able to solve that by using thread safe properties. `EditorPresenterTest` seems to be green. Fixing other tests in the meantime.

**Comment by saket** [comment]
iOS tests will now run as part of CI checks once https://github.com/saket/press/pull/26 is merged 🎉 

Update: done! Please rebase your branch?

File: shared/src/commonTest/kotlin/me/saket/press/shared/note/FakeNoteRepository.kt (hunk 1 above)
Comment (arkivanov) [reply shared/src/commonTest/kotlin/me/saket/press/shared/note/FakeNoteRepository.kt:44]: I'm still thinking about it but the reason is to freeze the downstream. Maybe it worth to hide it behind a meaningful extension function.

**Comment by arkivanov** [comment]
Now tests are failing because of freezes. Next commit will fix the production code.
```
2019-12-14T12:29:17.5156330Z [ RUN      ] me.saket.press.shared.editor.EditorPresenterTest.updating an existing note on exit when its content is non-blank
2019-12-14T12:29:17.5156730Z kotlin.native.concurrent.FreezingException: freezing of [Task(startMillis=0, periodMillis=null, task=$subscribe_2_2$<anonymous>_22$FUNCTION_REFERENCE$130@707229b8)] has failed, first blocker is me.saket.press.shared.editor.EditorPresenter@70721828
2019-12-14T12:29:17.5157150Z         at 0   test.kexe                           0x0000000102eac557 kfun:kotlin.Throwable.<init>(kotlin.String?)kotlin.Throwable + 87 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/Throwable.kt:22:37)
2019-12-14T12:29:17.5157800Z         at 1   test.kexe                           0x0000000102ea6105 kfun:kotlin.Exception.<init>(kotlin.String?)kotlin.Exception + 85 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/Exceptions.kt:23:44)
2019-12-14T12:29:17.5158120Z         at 2   test.kexe                           0x0000000102ea5d65 kfun:kotlin.RuntimeException.<init>(kotlin.String?)kotlin.RuntimeException + 85 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/Exceptions.kt:34:44)
2019-12-14T12:29:17.5158470Z         at 3   test.kexe                           0x0000000102ed5181 kfun:kotlin.native.concurrent.FreezingException.<init>(kotlin.Any;kotlin.Any)kotlin.native.concurrent.FreezingException + 641 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/concurrent/Freezing.kt:15:9)
2019-12-14T12:29:17.5158820Z         at 4   test.kexe                           0x0000000102ed551e ThrowFreezingException + 222 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/concurrent/Internal.kt:84:15)
2019-12-14T12:29:17.5159150Z         at 5   test.kexe                           0x00000001030f9f2b FreezeSubgraph + 2635
2019-12-14T12:29:17.5159460Z         at 6   test.kexe                           0x00000001031100ab Kotlin_Worker_freezeInternal + 27
2019-12-14T12:29:17.5159800Z         at 7   test.kexe                           0x0000000102ed4e6b kfun:kotlin.native.concurrent.freeze@#GENERIC.()Generic + 59 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/concurrent/Freezing.kt:33:5)
2019-12-14T12:29:17.5160170Z         at 8   test.kexe                           0x0000000102fedbb4 kfun:com.badoo.reaktive.utils.atomic.AtomicReference.freezeIfNeeded#internal + 276 (/Users/runner/runners/2.162.0/work/Reaktive/Reaktive/utils/src/nativeCommonMain/kotlin/com/badoo/reaktive/utils/atomic/AtomicReference.kt:22:13)
2019-12-14T12:29:17.5160540Z         at 9   test.kexe                           0x0000000102feda03 kfun:com.badoo.reaktive.utils.atomic.AtomicReference.compareAndSet(#GENERIC;#GENERIC)ValueType + 291 (/Users/runner/runners/2.162.0/work/Reaktive/Reaktive/utils/src/nativeCommonMain/kotlin/com/badoo/reaktive/utils/atomic/AtomicReference.kt:18:56)
2019-12-14T12:29:17.5161870Z         at 10  test.kexe                           0x000000010308a28d kfun:com.badoo.reaktive.test.scheduler.TestScheduler.Executor.addTask#internal + 861 (/Users/runner/runners/2.162.0/work/Reaktive/Reaktive/utils/src/commonMain/kotlin/com/badoo/reaktive/utils/atomic/AtomicReferenceExt.kt:9:15)
2019-12-14T12:29:17.5167300Z         at 11  test.kexe                           0x0000000103089177 kfun:com.badoo.reaktive.test.scheduler.TestScheduler.Executor.submit(kotlin.Long;kotlin.Function0<kotlin.Unit>) + 103 (/Users/runner/runners/2.162.0/work/Reaktive/Reaktive/reaktive-testing/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt:76:13)
2019-12-14T12:29:17.5167990Z         at 12  test.kexe                           0x000000010301252f kfun:com.badoo.reaktive.scheduler.Scheduler.Executor.submit$default(kotlin.Long;kotlin.Function0<kotlin.Unit>;kotlin.Int) + 351 (/Users/runner/runners/2.162.0/work/Reaktive/Reaktive/reaktive/src/commonMain/kotlin/com/badoo/reaktive/scheduler/Scheduler.kt:32:9)
2019-12-14T12:29:17.5168530Z         at 13  test.kexe                           0x0000000102ff55c9 kfun:com.badoo.reaktive.completable.<no name provided>_1_1.subscribe_2_2#internal.37 + 969 (/Users/runner/runners/2.162.0/work/Reaktive/Reaktive/reaktive/src/commonMain/kotlin/com/badoo/reaktive/completable/SubscribeOn.kt:9:0)
2019-12-14T12:29:17.5169070Z         at 14  test.kexe                           0x0000000102fee00f kfun:com.badoo.reaktive.base.subscribeSafe$reaktive@com.badoo.reaktive.base.Source<#GENERIC_com.badoo.reaktive.base.Observer_com.badoo.reaktive.base.ErrorCallback>.(#GENERIC_com.badoo.reaktive.base.Observer_com.badoo.reaktive.base.ErrorCallback)Generic + 271 (/Users/runner/runners/2.162.0/work/Reaktive/Reaktive/reaktive/src/commonMain/kotlin/com/badoo/reaktive/base/SubscribeSafe.kt:7:9)
2019-12-14T12:29:17.5170100Z         at 15  test.kexe                           0x0000000102ff43ea kfun:com.badoo.reaktive.completable.subscribe@com.badoo.reaktive.completable.Completable.(kotlin.Boolean;kotlin.Function1<com.badoo.reaktive.disposable.Disposable,kotlin.Unit>?;kotlin.Function1<kotlin.Throwable,kotlin.Unit>?;kotlin.Function0<kotlin.Unit>?)com.badoo.reaktive.disposable.Disposable + 1114 (/Users/runner/runners/2.162.0/work/Reaktive/Reaktive/reaktive/src/commonMain/kotlin/com/badoo/reaktive/completable/Subscribe.kt:33:12)
2019-12-14T12:29:17.5170680Z         at 16  test.kexe                           0x0000000102ff4736 kfun:com.badoo.reaktive.completable.subscribe$default@com.badoo.reaktive.completable.Completable.(kotlin.Boolean;kotlin.Function1<com.badoo.reaktive.disposable.Disposable,kotlin.Unit>?;kotlin.Function1<kotlin.Throwable,kotlin.Unit>?;kotlin.Function0<kotlin.Unit>?;kotlin.Int)com.badoo.reaktive.disposable.Disposable + 678 (/Users/runner/runners/2.162.0/work/Reaktive/Reaktive/reaktive/src/commonMain/kotlin/com/badoo/reaktive/completable/Subscribe.kt:11:1)
2019-12-14T12:29:17.5171230Z         at 17  test.kexe                           0x000000010306ed4d kfun:me.saket.press.shared.editor.EditorPresenter.saveEditorContentOnExit(kotlin.CharSequence) + 477 (/Users/runner/runners/2.162.0/work/press/press/shared/src/commonMain/kotlin/me/saket/press/shared/editor/EditorPresenter.kt:158:10)
2019-12-14T12:29:17.5172370Z         at 18  test.kexe                           0x000000010309d2f6 kfun:me.saket.press.shared.editor.EditorPresenterTest.updating an existing note on exit when its content is non-blank() + 822 (/Users/runner/runners/2.162.0/work/press/press/shared/src/commonTest/kotlin/me/saket/press/shared/editor/EditorPresenterTest.kt:115:15)
2019-12-14T12:29:17.5174070Z         at 19  test.kexe                           0x00000001030a0123 kfun:me.saket.press.shared.editor.$EditorPresenterTest$test$0.$updating an existing note on exit when its content is non-blank$FUNCTION_REFERENCE$5.invoke#internal + 67 (/Users/runner/runners/2.162.0/work/press/press/shared/src/commonTest/kotlin/me/saket/press/shared/editor/EditorPresenterTest.kt:108:9)
2019-12-14T12:29:17.5177480Z         at 20  test.kexe                           0x00000001030a0199 kfun:me.saket.press.shared.editor.$EditorPresenterTest$test$0.$updating an existing note on exit when its content is non-blank$FUNCTION_REFERENCE$5.$<bridge-UNNN>invoke(#GENERIC)#internal + 89 (/Users/runner/runners/2.162.0/work/press/press/shared/src/commonTest/kotlin/me/saket/press/shared/editor/EditorPresenterTest.kt:108:9)
2019-12-14T12:29:17.5177980Z         at 21  test.kexe                           0x0000000102f050d9 kfun:kotlin.native.internal.test.BaseClassSuite.TestCase.run() + 1289 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/internal/test/TestSuite.kt:85:26)
2019-12-14T12:29:17.5178370Z         at 22  test.kexe                           0x0000000102efee30 kfun:kotlin.native.internal.test.TestRunner.run#internal + 2416 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/internal/test/TestRunner.kt:229:0)
2019-12-14T12:29:17.5178730Z         at 23  test.kexe                           0x0000000102efffd2 kfun:kotlin.native.internal.test.TestRunner.runIteration#internal + 2930 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/internal/test/TestRunner.kt:252:0)
2019-12-14T12:29:17.5179080Z         at 24  test.kexe                           0x0000000102f008e2 kfun:kotlin.native.internal.test.TestRunner.run()ValueType + 962 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/internal/test/TestRunner.kt:270:0)
2019-12-14T12:29:17.5179450Z         at 25  test.kexe                           0x0000000102ef403b kfun:kotlin.native.internal.test.testLauncherEntryPoint(kotlin.Array<kotlin.String>)ValueType + 283 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/internal/test/Launcher.kt:22:53)
2019-12-14T12:29:17.5180030Z         at 26  test.kexe                           0x0000000102ef3ef7 kfun:kotlin.native.internal.test.main(kotlin.Array<kotlin.String>) + 55 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/internal/test/Launcher.kt:26:17)
2019-12-14T12:29:17.5180440Z         at 27  test.kexe                           0x0000000102ef418c Konan_start + 156 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/internal/test/Launcher.kt:25:1)
2019-12-14T12:29:17.5180790Z         at 28  test.kexe                           0x00000001030d168f Init_and_run_start + 111
2019-12-14T12:29:17.5181090Z         at 29  libdyld.dylib                       0x00007fff516ecd29 start + 1
```

**Comment by arkivanov** [comment]
The PR is ready for review. This is something you have to do if you want to keep your presenters and their dependencies not frozen.

**Comment by arkivanov** [comment]
Alright, after few additional fixes it appears that in this particular case it is quite difficult to keep presenters unfrozen. So I suggest to just design them in a freezable way. At the moment they look already freezable since they do not have any mutable state. So just keep this in mind. Closing the PR.

**Pull request closed without merging**